Create documentation for using excel ingestion and post upload tools on bastion server

pli888 opened the pull request:
# Pull request for issue: #1968 

This is a pull request for the following functionalities:

* Documentation for using excel ingestion and post upload processing tools on bastion server
* MD5 update functionality moved into files metadata console
* Removed reliance of S3 gigadb-dataset-metadata bucket by MD5 and file sizes update functionality 
* postUpload.sh script can perform whole post upload workflow
* Bats test for postUpload.sh
* Better error handling and reporting when file could not be updated with new file size or md5 value
* Works as curator or centos user on bastion server

## How to test?

### Dev environment

Perform steps in the **Test curator tools in dev environment** section in [docs/CURATOR_TOOLS_BASTION.md](https://github.com/gigascience/gigadb-website/blob/f280f712ccc8d8df4360408542766db38da547bb/docs/CURATOR_TOOLS_BASTION.md).

### Staging environment

Deploy your staging environment as usual:
```
$ ops/infrastructure/envs/staging
# Copy terraform files to staging environment
$ ../../../scripts/tf_init.sh --project gigascience/forks/<your>-gigadb-website --env staging --region <your-region> --ssh-key <path to your pem>

# Provision with Terraform 
$ terraform plan
$ terraform apply
$ terraform refresh

# Copy ansible files
../../../scripts/ansible_init.sh --env staging

# Provision webapp ec2 server with ansible
env TF_KEY_NAME=private_ip OBJC_DISABLE_INITIALIZE_FORK_SAFETY=YES ansible-playbook -i ../../inventories webapp_playbook.yml -e "gigadb_env=staging"
```

When you deploy your bastion server, the following command will restore your staging RDS to a specific date in the past which does not contain the dataset whose Excel file will be ingested for testing:
```
# Restore RDS to specific backupDate using databaseReset.sh with backup file sourced from pli888's EC2 FTP server
env OBJC_DISABLE_INITIALIZE_FORK_SAFETY=YES ansible-playbook -i ../../inventories bastion_playbook.yml -e "backupDate=20240101" -e "gigadb_env=staging"
```

The `gigadb-database-backup` S3 bucket already contains backup files that the above command will use:
* gigadb_gigascience-forks-rm-gigadb-website_staging_20240101.backup
* gigadb_gigascience-forks-kencho-gigadb-website_staging_20240101.backup

You will also need to create a curator user account on your bastion server, e.g.
```
# Create curator user peterl on bastion server
ansible-playbook -i ../../inventories users_playbook.yml -e "newuser=peterl" -e "credentials_csv_path=~/Desktop/credentials-peterl.csv" -e "gigadb_env=staging"
```

Run the Gitlab CI/CD pipeline for this feature branch.

To test Excel file ingestion and post-upload operations, perform the steps described in the Staging/Live section in [docs/CURATOR_TOOLS_BASTION.md](https://github.com/gigascience/gigadb-website/blob/f280f712ccc8d8df4360408542766db38da547bb/docs/CURATOR_TOOLS_BASTION.md).

These steps have been tested to work both as the `centos` user and a _curator_ user.

## How have functionalities been implemented?

### Remove use of S3 `gigadb-datasets-metadata` bucket

The md5.sh script does not copy doi.md5 and doi.filesizes into the S3 `gigadb-datasets-metadata` bucket anymore. Instead, these files are copied into `/var/share/gigadb/metadata/` directory in the bastion server. On your dev environment, `files-metadata-console/docker-compose.yml` maps the `/var/share/gigadb/metadata/` directory onto `files-metadata-console/tests/_data/var/share/gigadb/metadata/`.

The `DatasetFilesUpdater` component class in files-metadata-console has been updated to read file md5 values and file sizes from /var/share/gigadb/metadata/doi.md5 and /var/share/gigadb/metadata/doi.filesizes too. 

### Move tools into files metadata console

Md5 value update functionality used to be present FilesCommand in the protected directory. This functionality has been moved into files metadata console which involved adding updateMD5FileAttributes() function into DatasetFilesUpdater component class. The UpdateController class in files metadata console also has a new actionMd5Values() function which calls the updateMD5FileAttributes() function. The File class in gigadb/app/models has been updated with updateMd5Checksum($md5_value) function.

`filesMetaToDb.sh` and `postUpload.sh` moved from excel spreadsheet uploader tool to files metadata console because they are related to update file metadata, not ingestion of Excel files.

### Improved error handling in md5 value and file size update functionality

Error handling has been improved so that any file listed in doi.md5 and doi.filesizes but not found in database is reported in the console. Processing of the remainder of files in doi.md5 and doi.filesizes continues if a file not found error is encountered, where as before this did not happen.

A file size update bug has been fixed due to wrong delimiter used in `doi.filesizes` files which should be tab delimited. The `md5.sh` script replaces the single space separating file size and file name with a tab character.

### Fully working postUpload.sh

The postUpload.sh script is fully working, starting from readme file creation, copying readme file into user dropbox, calculation of md5 and file sizes and finally to the updating of md5 values and file size in the database.

## Any changes to automated tests?

Unit and functional tests in files metadata console have been updated. These tests can be executed as follows:
```
# Run unit tests
$ cd gigadb/app/tools/files-metadata-console
$ docker-compose run --rm files-metadata-console ./vendor/codeception/codeception/codecept run --debug tests/unit/
$ docker-compose run --rm files-metadata-console ./vendor/codeception/codeception/codecept run --debug tests/functional/
```

There is a new bats test for postUpload.sh:
```
$ bats postUpload
```


## Any changes to documentation?

Information about using command line tools on bastion server has been created in docs/CURATOR_TOOLS_BASTION.md. Curators can copy, paste this documentation into their Giga internet curator documentation page.

pli888 committed:
Write documentation for curators to use tools on bastion

pli888 committed:
Add more documentation

pli888 committed:
Fix image display in documentation

pli888 committed:
Update documentation using pull request 1954

pli888 committed:
Remove duplicated documentation

pli888 committed:
Add warning alerts

pli888 committed:
Stop copying metadata files to S3 bucket

pli888 committed:
Copy metadata files into a directory in /var

pli888 committed:
Add new bats test to check md5.sh from within container

pli888 committed:
Use doi.md5 from gigadb metadata directory to update md5 values

pli888 committed:
Add data for testing md5 update in dev environment

pli888 committed:
Create doi.filesizes with tab delimiters

The file size update tool in files metadata console requires
information in doi.filesizes to be separated by tabs.

pli888 committed:
Remove DOI from search string when searching for file to update

This was causing a bug stopping a file from being updated with
file size information if file was contained within a directory.

pli888 committed:
Throw useful exception message when file is not found

If a file in doi.filesizes is not found in database then an
exception with a useful message is thrown.

pli888 committed:
Change location of gigadb metadata directory on docker host

pli888 committed:
Update variable name

pli888 committed:
Remove unused constants

And correct command help documentation.

pli888 committed:
Provide data for unit and functional tests

pli888 committed:
Fix UpdateFileSizesWithNonExistentFile unit test

pli888 committed:
Move md5 update code into file metadata console tool

pli888 committed:
Add md5 file update unit tests for files metadata console

pli888 committed:
Add functional tests for updating md5 values

pli888 committed:
Tidy up try/catch blocks to catch exceptions

pli888 committed:
Enable file size updates to continue when a file cannot be found

pli888 committed:
Move success message into try block

Success message shoyld only be printed if no exception is caught.

pli888 committed:
Update functional test for checking file size update tool

pli888 committed:
Add new test to check when DOI cannot be found in database

pli888 committed:
Enable md5 update process to continue when file cannot be found

Sometimes a file is listed in doi.md5 but is not in database. A
message is displayed when this happens.

pli888 committed:
Delete files

Functionality for updating md5 values and file sizes is all
contained within files metadata console now.

pli888 committed:
Update how md5 value update tool is called

pli888 committed:
Fix typo

pli888 committed:
Move script into file metadata console

The filesMetaToDb.sh is a script that uses files metadata console.
If seems reasonable that it should be in this directory.

pli888 committed:
Move postUpload.sh into files metadata console

pli888 committed:
Add test dataset files

These files are required for dev testing curator tools workflow.

pli888 committed:
Map directories between host and container

This mapping is for accessing metadata files.

pli888 committed:
Mount metadata directory when running script on EC2

For accessing dataset metadata files

pli888 committed:
Add steps missing from curator workflow

pli888 committed:
Stop writing tool output into text files

As I've been managing dataset uploads for past few months, I do not
look back at these text files and moving them into the uploadDir
folder in the curator's home directory is difficult. For these
reasons, I have removed this code.

pli888 committed:
Tidy up status messages

pli888 committed:
Tidy up status messages

Removed some status messages. If there are problems when running
this script then tool being executed should generate an error
message.

pli888 committed:
Add test data into drop boxes

pli888 committed:
Remove old test data files

pli888 committed:
Add bats test for postUpload.sh

pli888 committed:
Update documentation for dev environment

pli888 committed:
Add check for readme file

pli888 committed:
Fix broken unit test

pli888 committed:
Fix md5 value update functional test

pli888 committed:
Fix update file size functional test

pli888 committed:
Update step to copy readme file into dropbox

Need to ensure readme file can be copied by centos and curator
users.

pli888 committed:
Update curator documentation for using tools on bastion

kencho51 commented on docs/CURATOR_TOOLS_BASTION.md:
````diff
@@ -0,0 +1,394 @@
+> [!TIP]
+> To continue with the remainder of this workflow, create a directory at /share/dropbox/user5 using the `centos` user. This user5 directory will act as an example user dropbox. Add the following two files into the user5 directory:
+```
+[peterl@ip-10-99-0-142 ~]$ mkdir /share/dropbox/user5
````
I got this error:
```
[lily@ip-10-99-0-95 ~]$ mkdir /share/dropbox/user5
mkdir: cannot create directory '/share/dropbox/user5': Permission denied
```

Which means that $user create dir in the efs.

The above path can be created as logged in as `centos`.

pli888 replied:
Now updated in md file.

kencho51 commented on gigadb/app/tools/files-metadata-console/controllers/UpdateController.php:
```diff
@@ -50,36 +48,42 @@ final class UpdateController extends Controller
+     * Console command for updating md5 values for all files in a given dataset
      *
-     * ./yii update/file-sizes --doi=100142
+     * ./yii update/file-sizes --doi=100039
```
**typo:** it should be `./yii update/md5-values --doi=100039`

pli888 replied:
Now updated.

kencho51 reviewed (commented):
Hi @pli888,

**praise:** I was able to follow steps in the new doc `CURATOR_TOOLS_BASTION.md`, and the doc is comprehensive and easy to follow.
**chore:** Should mention in the `CURATOR_TOOLS_BASTION.md` that the gitlab pipeline has been completed and passed before executing the tools in production, otherwise the some images, eg, production-files-metadata-console:staging is not updated and got errors like this:
```
[lily@ip-10-99-0-95 ~]$ sudo /usr/local/bin/filesMetaToDb 102498
Updating md5 checksum values as file attributes for 102498
Unknown command: update/md5-values
Caused by: Exception 'yii\base\InvalidRouteException' with message 'Unable to resolve the request: update/md5-values'

in /gigadb/app/tools/files-metadata-console/vendor/yiisoft/yii2/base/Controller.php:149

Stack trace:
#0 /gigadb/app/tools/files-metadata-console/vendor/yiisoft/yii2/console/Controller.php(180): yii\base\Controller->runAction('md5-values', Array)
#1 /gigadb/app/tools/files-metadata-console/vendor/yiisoft/yii2/base/Module.php(552): yii\console\Controller->runAction('md5-values', Array)
#2 /gigadb/app/tools/files-metadata-console/vendor/yiisoft/yii2/console/Application.php(180): yii\base\Module->runAction('update/md5-valu...', Array)
#3 /gigadb/app/tools/files-metadata-console/vendor/yiisoft/yii2/console/Application.php(147): yii\console\Application->runAction('update/md5-valu...', Array)
#4 /gigadb/app/tools/files-metadata-console/vendor/yiisoft/yii2/base/Application.php(384): yii\console\Application->handleRequest(Object(yii\console\Request))
#5 /gigadb/app/tools/files-metadata-console/yii(20): yii\base\Application->run()
#6 {main}

```
**praise:** All tools can be executed as expected in my dev and production staging.
**praise:** The workflow described in the overview.png is very useful and helpful.
**praise:** Great that `filesMetaToDb.sh` and `postUpload.sh` have been re-organized into files-metadata-console dir, which make the tools dir more structured.
**question(non-blocking):** Do we need to update the usage at here:  https://sites.google.com/d/1gaSPM1UIlCPWwgDzNbNQpVMAYf9kYZeO/p/1ecqjfPusM9yp4wovHZF70ypHvbcMcHRC/edit?pli=1&authuser=1
**question:** Is it necessary to create `/tests/var/share/gigadb/metadata` dir for the docker cmd, Can it atly be replaced by `tests/_data/dropbox/$user`? These two dir seem duplicated.
**praise:** Provided demo test data in the files-metadata-console, which is good for testing dev work

pli888 commented:
@kencho51 

>chore: Should mention in the CURATOR_TOOLS_BASTION.md that the gitlab pipeline has been completed and passed before executing the tools in production, otherwise the some images, eg, production-files-metadata-console:staging is not updated and got errors like this:

The documentation in `CURATOR_TOOLS_BASTION.md` was written for curators so I would prefer not to mention about gitlab pipeline has to be completed. From a curator's point of view, we assume the  gitlab pipeline has run to completion and the bastion server contains the latest version of the file metadata console.

>question(non-blocking): Do we need to update the usage at here: https://sites.google.com/d/1gaSPM1UIlCPWwgDzNbNQpVMAYf9kYZeO/p/1ecqjfPusM9yp4wovHZF70ypHvbcMcHRC/edit?pli=1&authuser=1

Yes we do. I will do this when this PR has been approved and merged.

>question: Is it necessary to create /tests/var/share/gigadb/metadata dir for the docker cmd, Can it atly be replaced by tests/_data/dropbox/$user? These two dir seem duplicated.

`/tests/_data/var/share/gigadb/metadata` and `tests/_data/dropbox/$user` are two different directories. `tests/_data/dropbox/$user` represents a user dropbox. `/tests/var/share/gigadb/metadata` is the directory where .filesizes and .md5 files are stored in the bastion server for files metadata console tool to use to update dataset file md5 value and filesizes in the database.

pli888 committed:
Update documentation

Dropbox directory needs to be created by centos user on bastion
server.

pli888 committed:
Update docblock for actionMd5Values function

only1chunts commented:
Question: 

> Not only are the 102498.md5 and 102498.filesizes files created in the user5 dropbox but they are also in the /var/share/gigadb/metadata/ directory on the bastion server. 

Does this mean that there are two different copies of the filesize and md5 values? If so, I thought we had previously agreed that there would only be 1 version of those files. There is a real danger that curators will not realise a second hidden copy exists somewhere else other than the userbox and therefore not know that any manual changes made in those files in the userbox will be useless.
Do we really need two copies of those files? 
If the scripts **absolutely** require those files in the /var/share/gigadb/metadata/ directory then we should not have a redundant copy in the userbox. However, if the scripts could use the userbox copy instead of the ones in /var/share/gigadb/metadata/ directory then we should adjust the scripts and remove the need for the hidden copy there.

pli888 commented:
As discussed in sprint status meeting, the database update of file metadata will default to using doi.filesizes and doi.md5 located in user dropbox directories.

kencho51 reviewed (approved):
Hi @pli888,

As all my requests have been addressed, I am happy to approve this PR.

rija reviewed (commented):
Hi @pli888,

**praise**: great to see the tools' code being moved into more sensible project directories
**praise**: the workflow diagram is very welcome and is useful
**question**: What does "Send gigadb.org link" on the diagram mean?

**issue** (non-blocking): it would better if the sections of the docs where exactly the legend under the step number in the diagram (1. datasetupload, 2. createreadme, ...)

**question**: who is the target audience for the `CURATOR_TOOLS_BASTION.md` ?
because there are references in there to steps/instructions on the dev environment, which curators don't need need to know,
and it will confuse them more.
**suggestion**: Have two docs, one in `docs/developers/` and one in `docs/curators/` with only the relevant information for each. 
The curators are free to copy the content of `docs/curators/` into the Giganet if they want to.
it looks like the dividing line is the first part that describe local testing, and the curator is the part that describe steps on AWS deployment

**typo**: `reqiires` instead of `requires`

**issue**: in the staging and live deployment, there is no need to describe how people login as everyone does it it's own way (the ssh keys are not always in the same place, some ssh from windows, some from mac)
**suggestion**: remove the parts where there are more than one way to do it.

**typo**: "A readme_.txt will appear in the `uploadDir` directory" 
the DOI placeholder doesn't render well on Github

**issue**: "To continue with the remainder of this workflow, create a directory at /share/dropbox/user5 using the `centos` user. This user5 directory will act as an example user dropbox. Add the following two files into the user5 directory:"
**suggestion**: That doesn't sounds like something for the curators to do, shouldn't be here if the docs is for curators, maybe put in the `docs/developers/` version of the docs?

**issue**: "The readme file in the uploadDir directory needs to be copied into the user dropbox:"
I can foresee curators making mistakes and/or grumbling on that task
**suggestion**: can we have all the tools created artefacts that matter to curators generated in the same directory that they run the script on ?
if they get the habit of running the tools from user drop boxes, they don't have any files shuffling to do. (our tools could even throw a warning 
when curators are trying to run them outside of user drop boxes)

**issue**: "Not only are the `102498.md5` and `102498.filesizes` files created in the user5 dropbox but they are also in the `/var/share/gigadb/metadata/` directory on the bastion server. Check this:"
as discussed at a past status meeting, not only the curators don't need to know about the latter location, I think there is going to be situations where the files at those locations might not be the same (ie: there may be exceptional occasion where thoses file in the user drop boxes need to be updated manually , and that can also happen accidentally). 
**suggestion**: if the script to update the database can use the files in the user dropbox as source, there is no risk of deviation between two locations and the curators will stop asking about the other location (which is the reason we wanted to change the current behaviour in the first place), and again this sits well with the curators running the tools from the user dropbox.

**nitpick**: the MD5 calculation can take a very long time, I wonder whether we can use Gum to display a spinner to continuously show feedback to the curators, so they don't have to wonder whether the tool is working or not
**chore**: I didn't' see an entry to CHANGELOG.md

**thought**: I wonder if it's best to hand over the tools gradually to the curators, to avoid curator brain overload and having to support them on all the tools at once. we could start by handing over (one by one?) the tools that constitutes the PostUpload workflow to start with. Now that there is a diagram, they shouldn't have problem figuring the place of each new tool in the grand scheme of things

**praise**: great to see sustained test coverage for both PHP and bash

pli888 committed:
Remove dev testing section and fix typos

pli888 committed:
Remove create user5 directory step

pli888 committed:
Use numbered steps in diagram as subtitles in documentation

pli888 committed:
Move documentation into curators directory

pli888 committed:
Create bastion tools documentation for developers

pli888 committed:
Create readme file in readmeFiles directory in dev environment

pli888 committed:
Allow readme file to be created in current working directory

Wherever this createReadme.sh is called, the readme file will be
created in current working directory when running in dev environment.

pli888 committed:
Create readme file in current working directory on staging and live

pli888 committed:
WIP Update logging to use /var/log directory

pli888 committed:
Write messages to log file in log directory in dev environment

Outputting a log file in uploadDir directory doesn't seem right.
Also, a single log file is used, not a new log file that is created
every time the readme tool is executed.

pli888 committed:
Fix rclone config file not found error

pli888 committed:
Update bats tests due to changes in createReadme script

Bats tests had to be updated because readme tool now outputs readme
file in user's working directory.

pli888 committed:
Fix broken functional tests

pli888 committed:
Update documentation to check it works on staging and live environments

pli888 committed:
Use metadata directory in tool's root directory

A metadata directory in the files metadata console's root directory
is used to access md5 and filesizes files.

pli888 committed:
Update unit test to check exception message if md5 file is not found

pli888 committed:
Remove steps to save metadata files in /var/share/gigadb/metadata/

pli888 committed:
Fix bats tests for md5 script

pli888 committed:
Use metadata files in current working directory by script

pli888 committed:
Delete metadata files

These files are not required because we are not using
/var/share/gigadb/metadata for storing metadata files anymore.

pli888 committed:
Update postUpload.sh to work in dev environment

These changes are required because this script is meant to be executed from a user dropbox directory.

pli888 committed:
Remove volume mapping as it is not required anymore

pli888 committed:
Delimit variables using braces

pli888 committed:
Remove docker volume definition

pli888 committed:
Assume we call scripts from a dropbox directory

pli888 committed:
WIP update documentation

pli888 committed:
Add gitignore file

pli888 committed:
WIP update documentation

pli888 committed:
Update check of current working directory in dev env

pli888 committed:
Update bats test

pli888 committed:
Tidy gitignore files

pli888 committed:
Fix path when calling createReadme script

pli888 committed:
Update curator documentation

pli888 committed:
Revert change in wasabi section

pli888 committed:
Add entry into CHANGELOG.md

pli888 committed:
Display spinner when calculating md5 values

pli888 committed:
Add checks to ensure scriptis are executed in user dropbox

pli888 committed:
Tidy up code

pli888 committed:
Fix correct directory checks

pli888 committed:
Add missing argument to function call

pli888 committed:
Install gum

pli888 commented:
> **question**: What does "Send gigadb.org link" on the diagram mean?

This is the step when I send the link to the dataset admin page to the curators after the ingestion process is complete. The new diagram replaces this step with `Go to dataset admin page on gigadb.org`

> issue (non-blocking): it would better if the sections of the docs where exactly the legend under the step number in the diagram (1. datasetupload, 2. createreadme, ...)

This has been implemented in gigadb-website/docs/curators/CURATOR_TOOLS_BASTION.md and gigadb-website/docs/developers/CURATOR_TOOLS_BASTION.md.

> suggestion: Have two docs, one in docs/developers/ and one in docs/curators/ with only the relevant information for each.

This has been implemented in gigadb-website/docs/curators/CURATOR_TOOLS_BASTION.md and gigadb-website/docs/developers/CURATOR_TOOLS_BASTION.md.

> issue: in the staging and live deployment, there is no need to describe how people login as everyone does it it's own way (the ssh keys are not always in the same place, some ssh from windows, some from mac) suggestion: remove the parts where there are more than one way to do it.

Done in gigadb-website/docs/curators/CURATOR_TOOLS_BASTION.md.

> issue: "To continue with the remainder of this workflow, create a directory at /share/dropbox/user5 using the centos user. This user5 directory will act as an example user dropbox. Add the following two files into the user5 directory:"

This is now only done in gigadb-website/docs/developers/CURATOR_TOOLS_BASTION.md for developer testing.

> issue: "The readme file in the uploadDir directory needs to be copied into the user dropbox:" I can foresee curators making mistakes and/or grumbling on that task suggestion: can we have all the tools created artefacts that matter to curators generated in the same directory that they run the script on ?

Now, all command-line tools work from a user dropbox directory. The workflow diagram has been updated to reflect this change. Command-line tools create artefacts, e.g. readme files, doi.filesizes, etc in a container directory which is mounted to the current working directory on the host. This pattern allows tools to work from a user dropbox directory.

> nitpick: the MD5 calculation can take a very long time, I wonder whether we can use Gum to display a spinner to continuously show feedback to the curators, so they don't have to wonder whether the tool is working or not

md5 calculation is now wrapped with a call to the gum tool to display a spinner when this process is running. This required installing gum in the files-metadata-console dockerfile. For staging and live environments, gum is already installed on the bastion server.

> thought: I wonder if it's best to hand over the tools gradually to the curators, to avoid curator brain overload and having to support them on all the tools at once. we could start by handing over (one by one?) the tools that constitutes the PostUpload workflow to start with. Now that there is a diagram, they shouldn't have problem figuring the place of each new tool in the grand scheme of things

Sure, we can discuss this with @onlyonechunts.

All typos have been fixed

pli888 commented:
# Testing

## Dev environment

To test on dev environment, follow instructions in [gigadb-website/docs/developers/CURATOR_TOOLS_BASTION.md](https://github.com/gigascience/gigadb-website/blob/f32be8ec5996bdc9d05e31eb680ae13a471e8d45/docs/developers/CURATOR_TOOLS_BASTION.md).

## Staging environment

To test on your staging server:
```
# Spin up your staging environment
../../../scripts/tf_init.sh --project gigascience/forks/your-gigadb-website --env staging --region your-region --ssh-key ~/.ssh/your-key.pem

# Provision with Terraform 
AWS_PROFILE=Name terraform plan
AWS_PROFILE=Name terraform apply
AWS_PROFILE=Name terraform refresh

# Copy ansible files
../../../scripts/ansible_init.sh --env staging

# Run webapp playbook
env TF_KEY_NAME=private_ip OBJC_DISABLE_INITIALIZE_FORK_SAFETY=YES ansible-playbook -i ../../inventories webapp_playbook.yml -e "gigadb_env=staging"

# Run bastion playbook
env OBJC_DISABLE_INITIALIZE_FORK_SAFETY=YES ansible-playbook -i ../../inventories bastion_playbook.yml -e "gigadb_env=staging"

# Create curator user on bastion server
ansible-playbook -i ../../inventories users_playbook.yml -e "newuser=peterl" -e "credentials_csv_path=~/Desktop/credentials-peterl.csv" -e "gigadb_env=staging"
```

Copy over files for testing onto staging bastion server:
```
$ pwd
/path/to/gigadb-website
$ scp -i ~/.ssh/your-key.pem -r gigadb/app/tools/files-metadata-console/tests/_data/dropbox/user5 centos@your-staging-ip:/share/dropbox
```

Check files have been copied onto staging server:
```
$ ssh -i ~/.ssh/your-key.pem centos@your-staging-ip
[centos@ip-10-99-0-120 ~]$ ls /share/dropbox/user5/
DLPFC_69_72_VNS_results.csv  E2_VNS_Ground_Truth.csv
```

Then follow [docs/curators/CURATOR_TOOLS_BASTION.md](https://github.com/gigascience/gigadb-website/blob/f32be8ec5996bdc9d05e31eb680ae13a471e8d45/docs/curators/CURATOR_TOOLS_BASTION.md) using your curator user account on bastion.

rija reviewed (commented):
Hi @pli888,

**praise**: the documentation for curators is clearer and more focused
**praise**: there is a developer docs for the curators tool

**issue**: I think the step 2 section (_Change directory to /share/dropbox/user directory_) as it is, is unnecessary  as it is neither a tool nor a process. I would have put the change directory command as a pre-amble/pre-requisite to all the subsequent sections that require the curator to be located in the user dropbox
Also the current step 2 in the diagram doesn't feel right as it seems to indicate that something is taken from the database and put in the user dropbox which is not correct.
**suggestion**: If there must be a step before readme creation, make step 2 section about preparation of the user dropbox: in curator docs, mention that the dropbox needs to be backed up as `user5.orig` before any manipulation and that the authors has received the FTP credentials and started uploading files. In the developer docs, mention that developers (for now) need to create the user drop box on the EFS and they do so with the `./make_dropbox.sh` script on the file server which also creates the FTP account for author to upload files

**typo**: grammar of `From this the user dropbox directory` in  _3. createReadme_ section

**question**: is there a reason for the curators to not use the parameters `--wasabi --apply --use-live-data` on `createReadme`?
if not, the code snippet should probably have them, otherwise the curators will forget to pass those parameters and wonder why their file is not public

**question**: when running `sudo /usr/local/bin/createReadme --doi 102498 --wasabi --apply` (both as centos and non-centos users), I have the following error in `/var/log/gigadb/readme.log`: 
```
2024/10/24 15:31:50 Failed to copy: NoCredentialProviders: no valid providers in chain. Deprecated.
	For verbose messaging see aws.Config.CredentialsChainVerboseErrors
```

But `/etc/aws/credentials`, `/home/centos/.aws/credentials` and `/home/athena/.aws/credentials` all have valid Wasabi credentials filled in.
Have you had that issue before? Am I missing a config ?(I think I followed the test instructions from the PR)

**issue**: I think the workflow as described in the docs and in the diagram is missing the final steps:
	7.  `transfer --wasabi --backup` that upload dataset files to Wasabi and back them up to S3 glacier
	8. Housekeeping of user dropboxes of published datasets (curator doc should mention curators needs delete user5 and user5.orig after ensuring that the files have been backed up to S3 glacier in previous step)
	9. Publish dataset and Mint a DOI 

**praise**: the `calculateChecksumSizes` works well, and the spinner does its job when huge files are involved

**typo**: `## 5. Run filesMetdaToDb to update file`
**suggestion**: the step 6 section should probably mention that curators can start creating **mockup pages** in the admin dashboard in order to preview the final dataset view page with the information that was added to the database in previous steps

**praise**: the `filesMetaToDb` command works well

**issue** (non-blocking):  when using `sudo`, curators have to use the full path to the command (e.g: `/usr/local/bin/calculateChecksumSizes`). It would be better I think if they can just type the command name, for clarity and speed, and  to further reduce opportunities for mistakes.
**suggestion**: This can be done by having in bastion playbook, after the `- name: Add curator to sudoers` task, a new task to add a `/etc/sudoers.d/89-defaults` file with content:
```
Defaults    secure_path = /sbin:/bin:/usr/sbin:/usr/bin:/usr/local/bin
```
that should override the defaults set in `/etc/sudoers`


(longer term, we probably shouldn't force curators to use `sudo`, but that would require a few changes in the bash scripts, not worth doing here)

pli888 committed:
Replace change directory step with alerts

pli888 committed:
Add step to create dropboxes using make_dropbox.sh

pli888 committed:
Add info about creating mockup page

pli888 committed:
Fix bug on bastion stopping readme to be copied into wasabi

pli888 committed:
Update documentation for uploading readme to wasabi

pli888 committed:
Add step to transfer files from drop box into Wasabi

pli888 committed:
Make files-env hidden

pli888 committed:
Add documentation to use transfer tool to copy files to wasabi

pli888 committed:
Add documentation for user dropbox house keeping

pli888 committed:
Update step 7 in documentation

pli888 committed:
Enable users from not having to specify full path to tools

pli888 committed:
Update diagram

pli888 commented:
>**issue**: I think the step 2 section (Change directory to /share/dropbox/user directory) as it is, is unnecessary as it is neither a tool nor a process. I would have put the change directory command as a pre-amble/pre-requisite to all the subsequent sections that require the curator to be located in the user dropbox. Also the current step 2 in the diagram doesn't feel right as it seems to indicate that something is taken from the database and put in the user dropbox which is not correct.

There is a banner at the beginning of each post-upload step which tells the user to ensure they are in a `/share/dropbox/user` directory. The diagram has also been updated accordingly.

>**typo**: grammar of From this the user dropbox directory in 3. createReadme section

This typo has been corrected.

>**question**: is there a reason for the curators to not use the parameters --wasabi --apply --use-live-data on createReadme?

The curator documentation has been updated in `Step 2. createReadme` to include how using these flags can enable the readme file to be uploaded into Wasabi.

>**question**: when running sudo /usr/local/bin/createReadme --doi 102498 --wasabi --apply (both as centos and non-centos users), I have the following error in /var/log/gigadb/readme.log:

I had the same problem on my staging server which is caused by rclone not being able to find the AWS credentials. To fix this, these variables have been added to the createReadme script:
```
export AWS_SHARED_CREDENTIALS_FILE=/home/centos/.aws/credentials
export AWS_PROFILE=wasabi-transfer
```

>**issue**: I think the workflow as described in the docs and in the diagram is missing the final steps:

>7.  transfer --wasabi --backup that upload dataset files to Wasabi and back them up to S3 glacier

The above has been added as `step 6. `transfer` - copy dataset files into Wasabi` into docs/curators/CURATOR_TOOLS_BASTION.md.

>8. Housekeeping of user dropboxes of published datasets (curator doc should mention curators needs delete user5 and user5.orig after ensuring that the files have been backed up to S3 glacier in previous step)

This has been added as step `7. Housekeeping of user dropboxes of published datasets` into docs/curators/CURATOR_TOOLS_BASTION.md.

>9. Publish dataset and Mint a DOI

I've not added this into curator documentation since I'm not completely sure what they do, after reading the curator docs on Google sites.

>typo: ## 5. Run filesMetdaToDb to update file

This has been fixed in step 4 in curator documentation.

>suggestion: the step 6 section should probably mention that curators can start creating mockup pages in the admin dashboard in order to preview the final dataset view page with the information that was added to the database in previous steps

This has been added into `5. Go to dataset admin page on gigadb.org` in curator documentation.

>issue (non-blocking): when using sudo, curators have to use the full path to the command (e.g: /usr/local/bin/calculateChecksumSizes). It would be better I think if they can just type the command name, for clarity and speed, and to further reduce opportunities for mistakes.

There is now a step in `ops/infrastructure/roles/bastion-users/tasks/main.yml` which creates a `/etc/sudoers.d/89-defaults` file so curators do not have to type full path to script for its execution. The documentation has been updated accordingly.

only1chunts commented on docs/curators/CURATOR_TOOLS_BASTION.md:
````diff
@@ -0,0 +1,249 @@
+[peterl@ip-10-99-0-95 user5]$ transfer --doi 102498 --sourcePath /share/dropbox/user5/ --backup --apply
+```
+
+After you have confirmed the files are safely stored in Wasabi and Glacier then the `user` and `user.orig` dropbox directories should be deleted to save storage space after the manuscript has been published.
````
it would be useful to have a hint as to how to check the Glacier for the stored files? And maybe a link to the rclone docs about checking Wasabi content)

pli888 replied:
I have mentioned checking `/var/log/gigadb/transfer.log` to see files have been backed up in S3 Glacier.

It would be better to allow curators to use rclone to list files in the S3 Glacier backup bucket. However, I think this will require cleaning up the rclone and aws configurations on bastion server and it would be better fixed and presented in a separate PR because this PR is already quite big and has been open for a while now.

only1chunts commented on docs/curators/CURATOR_TOOLS_BASTION.md:
```diff
@@ -0,0 +1,249 @@
+> [!TIP]
+> Take note of the number of changes made by the md5 and file size update tool. This number should be equal to the number of files listed in the metadata files.
+
+To ensure the postUpload script has worked, you should perform checks using the dataset, sample and file admin pages to see if dataset metadata are correctly stored in the database.
```
I dont think any of the postUpload scripts effect the sample table in any way.

pli888 replied:
I have removed mention of samples here now.

pli888 committed:
Update documentation

rija reviewed (approved):
Hi @pli888,

**praise**: all the tools now work for a non-centos user on AWS, and the `/usr/local/bin` prefix is no longer needed
**praise**: documentation for curators looks good
**issue** (non-blocking): the transfer command behave differently to the other tools by requiring  explicit `sourcePath`argument instead of operating on the current directory, but that can be implemented in a separate ticket. 
**praise**: by deleting FilesCommand, we've further remove dependency on legacy Yii 1


based on that I'm happy to approve

pli888 committed:
Delete FilesCommandCest because FilesCommand removed
